File: Data Structures & Algorithms/longest-repeating-substring-with-replacement/submission-12.cpp
class Solution {
public:
    int characterReplacement(string str, int k) {
        /*
        A -> 3
        B -> 2
        
        most freq occuring char - A
        while size of map (6) - size of most freq occur char (4) > k:
            move left ptr 
        */
        int longest_substr = 0;
        int left = 0;
        std::unordered_map<char, int> freq;

        for (int right = 0; right < str.size(); right++){
            freq[str[right]] += 1;

            while ((right - left + 1) - most_freq_char_occurence(freq) > k){
                freq[str[left++]] -= 1;
            }

            longest_substr = std::max(longest_substr, right - left + 1);
        }


        return longest_substr;
    }

private: 
    int most_freq_char_occurence(const std::unordered_map<char, int>& freq){
        int biggest_num = 0;

        for (auto& [_, val] : freq){
            biggest_num = std::max(biggest_num, val);
        }

        return biggest_num;
    }
};

File: Data Structures & Algorithms/longest-substring-without-duplicates/submission-12.py
class Solution:
    def lengthOfLongestSubstring(self, s: str) -> int:
        if len(s) == 0:
            return 0
            
        left, right = 0, 1
        substring = set(s[left])
        max_substring = 0
        while right < len(s):
            max_substring = max(max_substring, len(substring))
            while s[right] in substring:
                substring.remove(s[left])
                left += 1
            substring.add(s[right])
            right += 1
        return max(max_substring, len(substring))